### PR TITLE
2020: Implement table map construction

### PR DESCRIPTION
**Pull request by Manishearth**
Progress towards https://github.com/servo/servo/issues/27459

This PR does not do any table  layout whatsoever, nor does it even
construct boxes for table elements.

What this does is construct the CellMap, which makes it easy to find
which cells are occupying each slot of the table. This will be useful
when iterating over the table repeatedly during the layout algorithms.

### CODE REVIEW

File: components/layout_2020/table/construct.rs [hunk 1]
```diff
@@ -0,0 +1,509 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
```
Comment (SimonSapin) [inline components/layout_2020/table/construct.rs:4]: This is implementing https://html.spec.whatwg.org/multipage/tables.html#table-processing-model, right? Let’s have a link in a comment near the top of the file.

**Review by SimonSapin — COMMENTED** [review]
I still need to digest the core of the algorithm some more, but here are some comments so far.

If you need to rebase, please do that force-push separately from code changes so that I can find diffs for just the latter.

**Comment by SimonSapin** [comment]
I don’t have an opinion on `CellMap` v.s. `TableSlots` but let’s pick one name :)

File: components/layout_2020/table/construct.rs [hunk 2]
```diff
@@ -159,3 +305,41 @@ where
         // do something?
     }
 }
+
+impl<'a, 'builder, 'dom, Node> TableRowBuilder<'a, 'builder, Node>
+where
+    Node: NodeExt<'dom>,
+{
+    fn consume_rowspans(&mut self) {
+        loop {
```
Comment (pcwalton) [inline components/layout_2020/table/construct.rs:382]: nit: could this be `while let Some(...) = ...`?

File: components/layout_2020/table/construct.rs [hunk 3]
```diff
@@ -159,3 +305,41 @@ where
...
+            if let Some(span) = self.builder.incoming_rowspans.get_mut(self.current_x) {
+                if *span != 0 {
```
Comment (pcwalton) [inline components/layout_2020/table/construct.rs:385]: nit: could be `if *span == 0 { break; }` and then avoid some rightward drift (and would also be a bit clearer IMO)

File: components/layout_2020/table/construct.rs [hunk 4]
```diff
@@ -159,3 +305,41 @@ where
...
+                    *span -= 1;
+                    let previous = self
+                        .builder
+                        .slots
+                        .get_above(self.current_x)
+                        .expect("Cannot have nonzero incoming rowspan with no slot above");
+                    let new_slot = self.builder
+                        .slots.spanned_slot(self.current_x, self.builder.current_y(), previous)
+                        .expect("Nonzero incoming rowspan cannot occur without a cell spannign this slot");
```
Comment (pcwalton) [inline components/layout_2020/table/construct.rs]: typo: "spanning"

File: components/layout_2020/table/construct.rs [hunk 5]
```diff
@@ -4,16 +4,133 @@ use crate::dom_traversal::{
...
+    /// If (x, y) is spanned by an already resolved `spanner`, return offsets for a TableSlot::Spanned() for it
+    fn spanned_slot_single(
+        &self,
+        x: usize,
+        y: usize,
+        spanner: SlotAndLocation,
+    ) -> Option<(usize, usize)> {
+        if let TableSlot::Cell { width, height, .. } = spanner.slot {
+            if x >= spanner.x && y >= spanner.y && x < spanner.x + *width {
+                if *height == 0 || y < spanner.y + *height {
```
Comment (pcwalton) [inline components/layout_2020/table/construct.rs:99]: Why two nested `if`s here?

File: components/layout_2020/table/construct.rs [hunk 6]
```diff
@@ -310,36 +343,117 @@ impl<'a, 'builder, 'dom, Node> TableRowBuilder<'a, 'builder, Node>
...
+
+        if self.builder.incoming_rowspans.len() < current_x + colspan {
+            // make sure the incoming_rowspans table is large enough
+            // because we will be
+            self.builder
+                .incoming_rowspans
+                .resize(current_x + colspan, 0isize);
+        }
+
+        debug_assert!(
```
Comment (pcwalton) [inline components/layout_2020/table/construct.rs:450]: uber-nit: could be `debug_assert_eq!`

File: components/layout_2020/table/construct.rs [hunk 7]
```diff
@@ -310,36 +343,117 @@ impl<'a, 'builder, 'dom, Node> TableRowBuilder<'a, 'builder, Node>
...
+
+        let me = TableSlot::Cell {
+            cell: TableCellBox {},
+            width: colspan,
+            height: rowspan,
+        };
+
+        if self.builder.incoming_rowspans.len() < current_x + colspan {
+            // make sure the incoming_rowspans table is large enough
+            // because we will be
```
Comment (pcwalton) [inline components/layout_2020/table/construct.rs]: because we will be what?

File: components/layout_2020/table/construct.rs [hunk 8]
```diff
@@ -310,36 +343,117 @@ impl<'a, 'builder, 'dom, Node> TableRowBuilder<'a, 'builder, Node>
...
+        debug_assert!(
+            self.builder.incoming_rowspans[current_x] == 0,
+            "consume_rowspans must have been called before this!"
+        );
+
+        // if rowspan is zero, this is automatically negative and will stay negative
+        let outgoing_rowspan = rowspan as isize - 1;
+        self.builder.slots.push(me);
+        self.builder.incoming_rowspans[current_x] = outgoing_rowspan;
+        if colspan > 1 {
```
Comment (pcwalton) [inline components/layout_2020/table/construct.rs:459]: I don't think this `if` is necessary

File: components/layout_2020/table/construct.rs [hunk 9]
```diff
@@ -402,7 +437,7 @@ where
 
         if self.builder.incoming_rowspans.len() < current_x + colspan {
             // make sure the incoming_rowspans table is large enough
-            // because we will be
+            // because we will be writing to it
```
Comment (pcwalton) [inline components/layout_2020/table/construct.rs:444]: Oh, never mind, I saw you fixed up the comment here :)

File: components/layout_2020/table/construct.rs (hunk 2 above)
Comment (Manishearth) [reply components/layout_2020/table/construct.rs:382]: In later code I'm using the `current_x()` variable that's initialized outside of the if let.

File: components/layout_2020/table/construct.rs (hunk 3 above)
Comment (Manishearth) [reply components/layout_2020/table/construct.rs:385]: In later code that branch also gets complicated

File: components/layout_2020/table/construct.rs (hunk 5 above)
Comment (Manishearth) [reply components/layout_2020/table/construct.rs:99]: One is an && chain, the other is an || chain, mixing them looks confusing

**Comment by pcwalton** [comment]
This looks OK to me.

**Comment by Manishearth** [comment]
@bors-servo r=pcwalton

@SimonSapin if you want to continue where you left off, the unsquashed version of this is available on the `old-tables` branch on my fork

**Comment by Manishearth** [comment]
Wait no that's not an intermittent. Hm

**Comment by Manishearth** [comment]
```
+ ./mach filter-intermittents wpt-jsonsummary.log --log-intermittents intermittents.log --log-filteredsummary filtered-wpt-errorsummary.log --tracker-api default --reporter-api default
1 unexpected results that are NOT known-intermittents:
  ▶ Unexpected subtest result in /css/cssom/getComputedStyle-pseudo.html:
  │ FAIL [expected PASS] Resolution of width is correct for ::before and ::after pseudo-elements of display: contents elements
  │   → assert_equals: expected "50px" but got "50%"
  │ 
  │ @http://web-platform.test:8000/css/cssom/getComputedStyle-pseudo.html:76:18
  │ @http://web-platform.test:8000/css/cssom/getComputedStyle-pseudo.html:75:25
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ test@http://web-platform.test:8000/resources/testharness.js:535:30
  └ @http://web-platform.test:8000/css/cssom/getComputedStyle-pseudo.html:73:5
```

**Comment by Manishearth** [comment]
Ah, I broke display:contents. @pcwalton could you review the last commit?

**Comment by jdm** [comment]
```
Diff in /repo/components/layout_2020/style_ext.rs at line 504:
             },
             // This should not be a value of DisplayInside, but oh well
             // special-case display: contents because we still want it to work despite the early return
-            stylo::DisplayOutside::None if inside == stylo::DisplayInside::Contents => return Display::Contents,
(B+            stylo::DisplayOutside::None if inside == stylo::DisplayInside::Contents => {
(B+                return Display::Contents
(B+            },
(B             stylo::DisplayOutside::None => return Display::None,
         };
         let inside = match inside {
```

**Comment by mrobinson** [comment]
These changes have been rolled into #30799.

**Pull request closed without merging**